File: app/src/main/java/com/tr/hsyn/telefonrehberi/main/contact/data/ContactKey.java
package com.tr.hsyn.telefonrehberi.main.contact.data;


import com.tr.hsyn.contactdata.Account;
import com.tr.hsyn.contactdata.Contact;
import com.tr.hsyn.contactdata.ContactData;
import com.tr.hsyn.datakey.DataKey;
import com.tr.hsyn.label.Label;
import com.tr.hsyn.time.Time;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;
import java.util.Set;


/**
 * Defines the main keys for a contact.
 * The keys are used to access the contact data,
 * and with the keys, can be put any data in the contact.
 * Data privacy is done via {@link DataKey}.
 */
public interface ContactKey {
	
	/**
	 * Big pic (string)
	 */
	DataKey BIG_PIC      = DataKey.of(3, "bigPic");
	/**
	 * Numbers (string list)
	 */
	DataKey NUMBERS      = DataKey.of(4, "numbers");
	/**
	 * Emails -- string list
	 */
	DataKey EMAILS       = DataKey.of(5, "emails");
	/**
	 * Note -- string
	 */
	DataKey NOTE         = DataKey.of(6, "note");
	/**
	 * Events ({@link ContactData}) list
	 */
	DataKey EVENTS       = DataKey.of(7, "events");
	/**
	 * Groups -- string list
	 */
	DataKey GROUPS       = DataKey.of(8, "groups");
	/**
	 * Labels -- label list
	 */
	DataKey LABELS       = DataKey.of(9, "labels");
	/**
	 * Accounts -- string list
	 */
	DataKey ACCOUNTS     = DataKey.of(10, "accounts");
	/**
	 * Deleted date (long)
	 */
	DataKey DELETED_DATE = DataKey.of(11, "deletedDate");
	
	@Nullable
	static List<String> getNumbers(@NotNull Contact contact) {
		
		return contact.getData(NUMBERS);
	}
	
	@Nullable
	static List<String> getEmails(@NotNull Contact contact) {
		
		return contact.getData(EMAILS);
	}
	
	@Nullable
	static List<ContactData> getEvents(@NotNull Contact contact) {
		
		return contact.getData(EVENTS);
	}
	
	@Nullable
	static String getNote(@NotNull Contact contact) {
		
		return contact.getData(NOTE);
	}
	
	@Nullable
	static List<String> getGroups(@NotNull Contact contact) {
		
		return contact.getData(GROUPS);
	}
	
	@Nullable
	static List<String> getAccounts(@NotNull Contact contact) {
		
		return contact.getData(ACCOUNTS);
	}
	
	@Nullable
	static Set<Label> getLabels(@NotNull Contact contact) {
		
		return contact.getData(LABELS);
	}
	
	static boolean isDeleted(@NotNull Contact contact) {
		
		return contact.getLong(DELETED_DATE, 0L) != 0L;
	}
	
	static void setNumbers(@NotNull Contact contact, List<String> numbers) {
		
		contact.setData(NUMBERS, numbers);
	}
	
	static void setBigPic(@NotNull Contact contact, String pic) {
		
		contact.setData(BIG_PIC, pic);
	}
	
	static void setEmails(@NotNull Contact contact, List<String> emails) {
		
		contact.setData(EMAILS, emails);
	}
	
	static void setEvents(@NotNull Contact contact, List<ContactData> events) {
		
		contact.setData(EVENTS, events);
	}
	
	static void setLabels(@NotNull Contact contact, Set<Label> labels) {
		
		contact.setData(LABELS, labels);
	}
	
	static void setGroups(@NotNull Contact contact, List<String> groups) {
		
		contact.setData(GROUPS, groups);
	}
	
	static void setAccounts(@NotNull Contact contact, List<Account> accounts) {
		
		contact.setData(ACCOUNTS, accounts);
	}
	
	static void setNote(@NotNull Contact contact, String note) {
		
		contact.setData(NOTE, note);
	}
	
	static void setDeleted(@NotNull Contact contact) {
		
		contact.setData(DELETED_DATE, Time.now());
	}
	
}

File: app/src/main/java/com/tr/hsyn/telefonrehberi/main/contact/data/ContactKey.kt
package com.tr.hsyn.telefonrehberi.main.contact.data

import com.tr.hsyn.contactdata.ContactData
import tr.xyz.contact.Contact
import tr.xyz.kkey.Key

/**
 * Keys for contact.
 */
object ContactKey {
	
	//@off
	val PIC          : Key = Key(10, "pic")
	val BIG_PIC      : Key = Key(1, "bigPic")
	val NUMBERS      : Key = Key(2, "numbers")
	val EMAILS       : Key = Key(3, "emails")
	val NOTE         : Key = Key(4, "note")
	val EVENTS       : Key = Key(5, "events")
	val GROUPS       : Key = Key(6, "groups")
	val LABELS       : Key = Key(7, "labels")
	val ACCOUNTS     : Key = Key(8, "accounts")
	val DELETED_DATE : Key = Key(9, "deletedDate")
	//@on
}

// region Extensions for contact
// region var Contact.pic: String?
var Contact.pic: String?
	get() = this[ContactKey.PIC]
	set(value) {
		this[ContactKey.PIC] = value
	}
// endregion

// region var Contact.bigPic: String?
var Contact.bigPic: String?
	get() = this[ContactKey.BIG_PIC]
	set(value) {
		this[ContactKey.BIG_PIC] = value
	}
// endregion

// region var Contact.numbers: List<String>?
var Contact.numbers: List<String>?
	get() = this[ContactKey.NUMBERS]
	set(value) {
		this[ContactKey.NUMBERS] = value
	}
// endregion

// region var Contact.emails: List<String>?
var Contact.emails: List<String>?
	get() = this[ContactKey.EMAILS]
	set(value) {
		this[ContactKey.EMAILS] = value
	}
// endregion

// region var Contact.note: String?
var Contact.note: String?
	get() = this[ContactKey.NOTE]
	set(value) {
		this[ContactKey.NOTE] = value
	}
// endregion

// region var Contact.events: List<ContactData>?
var Contact.events: List<ContactData>?
	get() = this[ContactKey.EVENTS]
	set(value) {
		this[ContactKey.EVENTS] = value
	}
// endregion

// region var Contact.groups: List<String>?
var Contact.groups: List<String>?
	get() = this[ContactKey.GROUPS]
	set(value) {
		this[ContactKey.GROUPS] = value
	}
// endregion

// region var Contact.labels: List<String>?
var Contact.labels: List<String>?
	get() = this[ContactKey.LABELS]
	set(value) {
		this[ContactKey.LABELS] = value
	}
// endregion

// region var Contact.accounts: List<String>?
var Contact.accounts: List<String>?
	get() = this[ContactKey.ACCOUNTS]
	set(value) {
		this[ContactKey.ACCOUNTS] = value
	}
// endregion

// region var Contact.deletedDate: Long?
var Contact.deletedDate: Long?
	get() = this[ContactKey.DELETED_DATE]
	set(value) {
		this[ContactKey.DELETED_DATE] = value
	}
// endregion
// endregion

